src/Physics/HadronTransport/INukeUtils2025.cxx

```diff
@@ -0,0 +1,1934 @@
+
+  p->SetPosition(x4new);
+}
+
+
+//___________________________________________________________________________
+// Method to handle compound nucleus considerations, preequilibrium
+//    and equilibrium
+//    Alex Bell -- 6/17/2008
+void genie::utils::intranuke2025::PreEquilibrium(
+  GHepRecord * ev, GHepParticle * p,
+  int &RemnA, int &RemnZ, TLorentzVector &RemnP4,
+  bool /* DoFermi */, double /* FermiFac */,
+  const NuclearModelI* /* Nuclmodel */, double NucRmvE, EINukeMode mode)
+{
+
+#ifdef __GENIE_LOW_LEVEL_MESG_ENABLED__
+  LOG("INukeUtils", pDEBUG)
+    << "PreEquilibrium() is invoked for a : " << p->Name()
+    << " whose kinetic energy is : " << p->KinE();
+#endif
+
+  // Random number generator
+  RandomGen * rnd = RandomGen::Instance();
+  //unused PDGLibrary * pLib = PDGLibrary::Instance();
+
+  bool allow_dup = true;
+  PDGCodeList list(allow_dup); // list of final state particles
+
+  double ppcnt = (double) RemnZ / (double) RemnA; // % of protons left
+
+  // figure out the final state conditions
+
+  if(p->Pdg()==kPdgProton) list.push_back(kPdgProton);
+  else if(p->Pdg()==kPdgNeutron) list.push_back(kPdgNeutron);
+
+  for(int i=0;i<3;i++)
+    {
+      if(rnd->RndFsi().Rndm()<ppcnt)
+        {
+          list.push_back(kPdgProton);
+          RemnZ--;
+        }
+      else list.push_back(kPdgNeutron);
+
+      RemnA--;
+
+      ppcnt = (double) RemnZ / (double) RemnA;
+    }
+
+  // Add the fermi energy of the three nucleons to the phase space
+  /*
+  if(DoFermi)
+    {
+      Target target(ev->TargetNucleus()->Pdg());
+      TVector3 pBuf = p->P4()->Vect();
+      double mBuf = p->Mass();
+      double eBuf = TMath::Sqrt(pBuf.Mag2() + mBuf*mBuf);
+      TLorentzVector tSum(pBuf,eBuf);
+      double mSum = 0.0;
+      vector<int>::const_iterator pdg_iter;
+      for(pdg_iter=++(list.begin());pdg_iter!=list.end();++pdg_iter)
+        {
+          target.SetHitNucPdg(*pdg_iter);
+          Nuclmodel->GenerateNucleon(target);
+          mBuf = pLib->Find(*pdg_iter)->Mass();
+          mSum += mBuf;
+          pBuf = FermiFac * Nuclmodel->Momentum3();
+          eBuf = TMath::Sqrt(pBuf.Mag2() + mBuf*mBuf);
+          tSum += TLorentzVector(pBuf,eBuf);
+          RemnP4 -= TLorentzVector(pBuf,eBuf-mBuf);
+        }
+      TLorentzVector dP4 = tSum + TLorentzVector(TVector3(0,0,0),-mSum);
+      p->SetMomentum(dP4);
+      }
+  */
+  // do the phase space decay & save all f/s particles to the event record
+  bool success = genie::utils::intranuke2025::PhaseSpaceDecay(ev,p,list,RemnP4,NucRmvE,mode);
+  if(success)  LOG("INukeUtils2025",pINFO) << "Successful phase space decay for pre-equilibrium nucleus FSI event";
+  else
+    {
+      exceptions::INukeException exception;
+      exception.SetReason("Phase space generation of pre-equilibrium nucleus final state failed, details above");
+      throw exception;
+    }
+
+  int p_loc = 0;
+  while(p_loc<ev->GetEntries())
+    {
+      GHepParticle * p_ref = ev->Particle(p_loc);
```

Comment:
Maybe a word here about the intent of this nested if/then/elses -- I _think_ it is to find the entry in the event `ev` that corresponds to the GHepParticle `p` that was passed in.  But clarifying that intent would make understanding the code easier.